### ISSUE TITLE
Authroization Code Flow for OpenId

### ISSUE THREAD

**Issue by dagostinelli** [xamarin/Xamarin.Auth#230]
Does Xamarin.Auth support the Authorization Code Flow?  

Regarding: https://security.stackexchange.com/questions/129928/oidc-flow-for-spa-and-restful-api.

**Comment by moljac** [comment]
> Does Xamarin.Auth support the Authorization Code Flow?

Yes - for OAuth2

... but your link is for OpenId...

So after short analysis.

Did you mean:

https://openid.net/specs/openid-connect-core-1_0.html#CodeFlowAuth

???

Authorization Code Flow for OpenID from that spec does not use `client_secret`, so it is more like

So, OAuth2Authenticator ctors will not be suitable for that right now, but I'm working on adding API for custom parameters which is nothing else but addition (superset) to OAuth2 for OpenID.

If this is related to OpenId pls change issue name to: "Authroization Code Flow for OpenId"

**Comment by moljac** [comment]
Which OpenId service providers use Authorization Code Flow?
(for testing purposes)

**Comment by dagostinelli** [comment]
I apologize if I've used the terms incorrectly.  I think of it like this:  Authorization Code flow was the default and implicit grant was sugar on top to make this go a little easier if all you need to do is directly get the user's profile every time.

All this means to me is that you hit the authorization url that the providers give you but don't exchange the code for an access token / id_token.  (Google and GitHub for me)

I intend to call something on my rest end point (https://myapi.com/authorize?code=<this is the code that google sends back>&idp=google) and then get back my own JWT bearer token that is specific to my rest api and signed by my rest api.  

Inside the REST API on the first call, it goes out to Google (just the first time) checks that the code is good, exchanges it for an access_token (and an id_token only if necessary).

At least this is how I'm intending to do it ([just like how that article says](https://security.stackexchange.com/questions/129928/oidc-flow-for-spa-and-restful-api))

It feels appropriate.  Am I missing anything important?

**Comment by dagostinelli** [comment]
After I wrote you, I went to implement this, at least in part, and ran into some problems with Google's endpoint sending back [invalid_grant](https://blog.timekit.io/google-oauth-invalid-grant-nightmare-and-how-to-fix-it-9f4efaf1da35). I'm having this problem while attempting to exchange the code for an access_token on my server side.  This seems odd to me because I thought that's how Auth flow works.  

After some reading, I came across these links that say OpenId2 is deprecated.
https://developers.google.com/identity/protocols/OpenID2Migration
https://developers.google.com/identity/protocols/OpenID2

Regardless of these standards and such, it sounds like Google, at least, has a very specific way they want us to use their service.  I'll close this issue now since it seems irrelevant.
